Use private heap for 'chip simple memory allocator'

andy31415 opened the pull request:
#### Problem
Chip simple memory allocation was not as simple after all and is optimized only for large blocks.

 #### Summary of Changes
- replace SimpleAlloc with 'PrivateHeap' instead (a linked list heap allocator)
- Remove the 'isLongTerm' parameter to platform memory alloc (has no meaning anymore)
- add implementation and unit tests for 'realloc' functionality.

msandstedt commented on src/lib/support/PrivateHeap.h:
```diff
@@ -38,6 +38,12 @@ void * PrivateHeapAlloc(void * heap, size_t size);
+// Reallocate a block to a new size
+// Special arguments:
+//    - null ptr value represents an Alloc
+//    - zero size represents a Free
```
Are we sure we want to support these special usages when standard implementations do not?

andy31415 replied:
I took these from the realloc documentation from Linux:

```
If ptr is NULL, then the call is equivalent to malloc(size), for all values of size; if size is equal to zero, and ptr is not NULL,
       then the call is equivalent to free(ptr).Y
       ```

msandstedt replied:
Sorry, I should have said that nat all standard implementations support these.

This is fine.  Not a big deal.  It's clearly defined in the comments.  Looks good.

msandstedt commented on src/lib/support/CHIPMem-Simple.cpp:
```diff
@@ -0,0 +1,92 @@
+
+void MemoryShutdown()
+{
+    gPrivateHeap = nullptr;
```
What is the purpose of shutdown when there still can be outstanding allocations?

andy31415 replied:
I am unclear on the purpose of shutdown in all cases :/ once you allocate heap it seems reasonable that you never clear it. Have this because old code has it. not very clear to me.

msandstedt commented:
Is there a plan to add locking?  `chip_config_memory_management=simple` builds aren't currently thread safe.

msandstedt commented:
Since this is the commit that activates the heap allocator, I think it is fair game to at least start discussions on the allocator itself here.  I'm happy to open new issues if need be.

In addition to the lack of locking (admittedly not new with the heap allocator), here are some other things that I see reading through `src/lib/support/PrivateHeap.cpp`.  It is only by accident that it achieves `kPrivateHeapAllocationAlignment`.  Were the header to have a smaller alignment and/or be packed, this would break.  That's dangerous.

The header is also oddly large.  The checksum is very weak, but yet occupies 4 bytes.  Notably, it will evaluate as uncorrupted for a header structure that's all zero.  That seems like a big miss given its intended purpose.

I know this is not newly-written code, but I think we should evaluate it carefully.

andy31415 commented:
> Is there a plan to add locking? `chip_config_memory_management=simple` builds aren't currently thread safe.

The thing that I replaced did not have locking, however I totally agree that heap is expected to be thread safe. I added some locking constructs. 

Not ISR safe, however I believe that maybe not allowing heap operations in ISRs is an acceptable limitation.

andy31415 commented:
> Since this is the commit that activates the heap allocator, I think it is fair game to at least start discussions on the allocator itself here. I'm happy to open new issues if need be.
> 
> In addition to the lack of locking (admittedly not new with the heap allocator), here are some other things that I see reading through `src/lib/support/PrivateHeap.cpp`. It is only by accident that it achieves `kPrivateHeapAllocationAlignment`. Were the header to have a smaller alignment and/or be packed, this would break. That's dangerous.
> 
> The header is also oddly large. The checksum is very weak, but yet occupies 4 bytes. Notably, it will evaluate as uncorrupted for a header structure that's all zero. That seems like a big miss given its intended purpose.
> 
> I know this is not newly-written code, but I think we should evaluate it carefully.

PrivateHeap has a static assert that its size is matching alignment, so that should be ok. I believe this is why we have the checksum value so large even though I totally agree it is very weak and could be smaller.

Totally agree on the 'all zero' issue and that checksums could be better. Maybe we should have a debug vs non-debug implementation to more easily capture coruption issues (stronger checksums)? Likely we will not reach the level a non-embedded platform can provide though (like asan and such).

Happy to improve on this.

andy31415 commented:
I tried this out in esp32 by forcing mem alloc simple. Found a few things:
  - fixed several compile issues (yay!)
  - esp32 apparently is built as 'CHIP_NO_LOCKING', so apparently we have to support that mode.

bzbarsky-apple commented on src/lib/support/CHIPMem-Simple.cpp:
```diff
@@ -0,0 +1,130 @@
+{
+    size_t total = num * size;
+
+    // check is for multiplication overflow
```
```suggestion
    // check for multiplication overflow
```

bzbarsky-apple commented on src/lib/support/tests/TestPrivateHeap.cpp:
```diff
@@ -264,6 +265,74 @@ void BackwardFreeAndRealloc(nlTestSuite * inSuite, void * inContext)
+    p1 = allocator.HeapRealloc(p2, 16); // reallocation does not change block size
+    NL_TEST_ASSERT(inSuite, p1 == p2);
+
+    p3 = allocator.HeapAlloc(48); // Heap still insufficient heap for this
```
```suggestion
    p3 = allocator.HeapAlloc(48); // Still insufficient heap for this
```

mspang commented on src/lib/support/PrivateHeap.cpp:
```diff
@@ -267,6 +269,42 @@ extern "C" void PrivateHeapFree(void * ptr)
+    void * largerCopy = PrivateHeapAlloc(heap, size);
+    if (largerCopy == nullptr)
+    {
+        PrivateHeapFree(ptr);
```
I think this is free not permitted?

andy31415 replied:
You are right - did not closely read realloc return documentation.

Updated.
